src/java/htsjdk/samtools/util/AsyncBufferedIterator.java

```diff
@@ -0,0 +1,209 @@
+
+/**
+ * Iterator that uses a background thread to perform read-ahead to improve
+ * throughput at the expense of increased latency.
+ * 
+ * @author Daniel Cameron
+ *
+ */
+public class AsyncBufferedIterator<T> implements CloseableIterator<T> {
+    private static final Log log = Log.getInstance(AsyncBufferedIterator.class);
+    /** Count of threads created */
+    private static AtomicInteger threadsCreated = new AtomicInteger(0);
+    private final Thread reader;
+    private final ReaderRunnable readerRunnable;
+    private final AtomicReference<Throwable> ex = new AtomicReference<Throwable>(null);
+    private final Iterator<T> underlying;
+    private final BlockingQueue<List<T>> buffer;
+    private final int bufferSize;
+    private boolean closeCalled = false;
+    /**
+     * Current buffer of records returned from the background thread. An empty
+     * list indicates the background thread has encountered the end of the
+     * underlying stream.
+     */
+    private List<T> currentList = null;
+    private Iterator<T> currentBuffer = null;
+
+    /**
+     * Creates a new iterator that traverses the given iterator on a background
+     * thread
+     * 
+     * @param iterator iterator to traverse
+     * @param bufferCount number of read-ahead buffers
+     * @param bufferSize size of each read-ahead buffer. A larger size will increase both throughput and latency.
+     */
+    public AsyncBufferedIterator(final Iterator<T> iterator, final int bufferCount, final int bufferSize) {
+        this(iterator, null, bufferCount, bufferSize);
+    }
+
+    /**
+     * Creates a new iterator that traverses the given iterator on a background
+     * thread
+     * 
+     * @param iterator iterator to traverse
+     * @param threadName background thread name
+     * @param bufferCount number of read-ahead buffers
+     * @param bufferSize size of each read-ahead buffer. A larger size will increase both throughput and latency.
+     */
+    public AsyncBufferedIterator(final Iterator<T> iterator, final String threadName, final int bufferCount, final int bufferSize) {
+        if (iterator == null) throw new IllegalArgumentException();
+        if (bufferCount <= 0) throw new IllegalArgumentException("Must use at least 1 buffer.");
+        if (bufferSize <= 0) throw new IllegalArgumentException("Buffer size must be at least 1 record.");
+        this.underlying = iterator;
+        this.buffer = new ArrayBlockingQueue<List<T>>(bufferCount);
+        this.bufferSize = bufferSize;
+        this.readerRunnable = new ReaderRunnable();
+        this.reader = new Thread(readerRunnable, threadName == null ? getThreadNamePrefix() + threadsCreated.incrementAndGet() : threadName);
+        this.reader.setDaemon(true);
+        log.debug("Starting thread " + this.reader.getName());
+        this.reader.start();
+    }
+
+    protected String getThreadNamePrefix() {
+        return "AsyncBufferedIterator";
```

Comment:
how about `AsyncBufferedIterator.class.getSimpleName()` instead (for the ease of refactoring etc)?